### DISCUSSION TITLE
Source generators guide/help

### DISCUSSION THREAD

**Discussion by Cybrosys** [CommunityToolkit/Maui#1578] [Q&A] [answered]
Hello!

I've written a simple source generator for doing DI service registration by using attributes. I am however having an issue where I can't deploy to iOS (hot reload) unless I add this package reference in my app's csproj file:
`<PackageReference Include="Microsoft.CodeAnalysis.CSharp" Version="4.8.0" />`

To me that is wrong, but unless I do it I get this error:
```
2023-11-29 10:30:28.104 Xamarin.PreBuilt.iOS[14916:12722053] Could not resolve assembly Microsoft.CodeAnalysis, Version=4.8.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35. Details: Could not load file or assembly '/var/mobile/Containers/Data/Application/F4524867-5DDC-4964-8BC4-FFF8AEE59532/Documents/StoreApp.content/Microsoft.CodeAnalysis.dll' or one of its dependencies.
```

I don't have this issue using your source generator nor any other source generator, this is only an issue with my own source generator project that my app is referencing.

I am referencing my SG project like this:
`<ProjectReference Include="<path to .csproj>" OutputItemType="Analyzer" />`

I am not using the attribute `ReferenceOutputAssembly="false"` as my SG project has the attributes that my SG is looking for.

I tried creating a separate project that only contained the attributes, however, the source generator stopped working then for some reason.

So yeah, I would really like some guide/help as to how to create a SG for MAUI that works, since the Community Toolkit has one and several others work. Is this perhaps an issue with Hot reload and that my SG projected is referenced instead of having it as a NuGet package?

**Comment by pictos** [comment]
@Cybrosys can you push your code to github and share the link? That will be easier to find what's going on.

Did you add the `PrivateAssets` property on your `PackageReference`? Like [here](https://github.com/CommunityToolkit/Maui/blob/main/src/CommunityToolkit.Maui.SourceGenerators/CommunityToolkit.Maui.SourceGenerators.csproj#L13-L14)

And I recommend you to try setting the `ReferenceOutputAssembly="false"` on your app project as well

**Reply by Cybrosys** [reply]
Of course, here's an app and a source generator showing the usage of the attributes:
https://github.com/Cybrosys/maui-repros/tree/main/MauiAppSourceGenerators

The source generator project has attributes that the app decorates its classes with, so that is why I can't use `ReferenceOutputAssembly="false"`. I tried moving the attributes to a separate project, but then the source generator stopped working for some reason.

I've added a `ICounterService` to `MauiProgram.cs` and added the `[Transient]` attribute on the `MainPage` and `[Transient<ICounterService>]` on the `CounterService` implementation.

The source generator generates an extension method on MauiBuilder (`.UseSourceGeneratedServiceRegistration()`).

I have no issues running this on an Android emulator, but on iOS HR blows up.

**Reply by pictos** [reply, marked as answer]
> The source generator project has attributes that the app decorates its classes with, so that is why I can't use ReferenceOutputAssembly="false". I tried moving the attributes to a separate project, but then the source generator stopped working for some reason.


I used this approach for SG with attributes and it works. You can see here
https://github.com/CommunityToolkit/Maui/pull/1321 

look for `src/CommunityToolkit.Maui/CommunityToolkit.Maui.csproj` 

I'll be able to look your code sometime next week

**Reply by Cybrosys** [reply]
@pictos thank you.

I looked it over and got it working. 

- I moved the attributes into their own separate project.
- I updated the source generator to use const strings for the full names of the attributes instead of using `typeof(<Attribute>).FullName`. 
- I updated the main project to reference both projects:
```
<ProjectReference Include="..\MauiApp1.DependencyInjection\MauiApp1.DependencyInjection.csproj" />
<ProjectReference Include="..\MauiApp1.DependencyInjection.SourceGenerators\MauiApp1.DependencyInjection.SourceGenerators.csproj" OutputItemType="Analyzer" ReferenceOutputAssembly="false" />
```

I've updated the repro with the changes.

I don't know if how I'm doing things in the source generator is good or not, source generation is something I've just started looking at.

**Reply by pictos** [reply]
Sadly there's no good docs around SG, make sure you don't break the incremental behavior, if you are using an incremental source generator, if can do that you're more than good to go (:

**Reply by IeuanWalker** [reply]
@pictos do you mind taken a look at my SG having a similar issue but cant get my head around it. I've looked at your PR and it looks very similar to what im doing.

The SG seamed to work and finding the classes/ attributes it needs but the `NameArguemnts` is always empty - https://github.com/IeuanWalker/Hangfire.RecurringJob.Generator/blob/21547628cd9f2d84aba82a25bac40700b070201d/Scr/RecuringJobGenerator.cs#L64C27-L64C52

I've posted the issue on StackOverflow but not getting any help yet lol - https://stackoverflow.com/questions/77609821/sourcegenerator-attribute-namedarguements-is-empty

**Reply by pictos** [reply]
@IeuanWalker in order to extract information passed through ctor, you need to use another approach, as you can see [here](https://github.com/CommunityToolkit/Maui/blob/7711bb7bf299457551f7369bc71d9cd7e33bfd56/src/CommunityToolkit.Maui.SourceGenerators.Internal/BindablePropertyAttributeSourceGenerator.cs#L172) I used a different approach when the info is passed on ctor and another way for optional properties. You can find all my extension methods [here](https://github.com/CommunityToolkit/Maui/blob/7711bb7bf299457551f7369bc71d9cd7e33bfd56/src/CommunityToolkit.Maui.SourceGenerators.Internal/Helpers/AttributeExtensions.cs)

**Reply by IeuanWalker** [reply]
@pictos pretty sure thats what im doing, these are the properties im trying to get - 
![image](https://github.com/CommunityToolkit/Maui/assets/6544051/3815d76c-5c2f-42d5-869b-c0969c563082)

And this is what im doing, which looks like the extention method helper you have - 
![image](https://github.com/CommunityToolkit/Maui/assets/6544051/e6758510-07ad-4325-a3b2-28a369e485d0)

**Reply by pictos** [reply]
I saw that you're using the `this` [overload for ctors](https://github.com/IeuanWalker/Hangfire.RecurringJob.Generator/blob/21547628cd9f2d84aba82a25bac40700b070201d/Scr/Attributes/RecurringJobAttribute.cs#L57) and populating some properties using it, I don't believe that you can grab information from it on compilation-time, since it's a more complex thing, at least I don't know how to do it and didn't find any sample where that approach is used. I recommend you to rework your attribute in order to make it works as you expect.

**Reply by pictos** [reply]
@IeuanWalker update, I found a sample where the `this` chain for [ctor is used](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/GeneratedRegexAttribute.cs), the thing is that there's no default value used, so I believe that's the issue, maybe you need to handle defaults inside your generator. This makes sense, because when you're running a SourceGenerator, you're reading the code the user typed, so if there's nothing there it will return `null`, and you need to handle the defaults... Now if were reading your attribute on runtime you should be able to see the defaults set.

**Reply by IeuanWalker** [reply]
@pictos ahh ok, that makes sense, been driving me crazy for a while lol

So i could just leave the attribute as is and just replicate the default value logic manually depending on how many constructor parms are passed through.

Thanks for taking a look :)

**Reply by pictos** [reply]
Glad that I could help you

**Reply by IeuanWalker** [reply]
working like a charm now - 
![image](https://github.com/CommunityToolkit/Maui/assets/6544051/625a054c-023a-4ed4-8c44-e1ed0c8fb51e)

**Comment by CliffAgius** [comment]
Closed as Answered
